File: src/main/java/work/xeltica/craft/core/stores/CounterStore.java
package work.xeltica.craft.core.stores;

import org.bukkit.Location;
import org.bukkit.configuration.serialization.ConfigurationSerialization;
import work.xeltica.craft.core.models.CounterData;
import work.xeltica.craft.core.models.PlayerDataKey;
import work.xeltica.craft.core.utils.Config;

import javax.annotation.Nullable;
import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 時間計測カウンターの情報を管理します。
 */
public class CounterStore {
    public CounterStore() {
        ConfigurationSerialization.registerClass(CounterData.class, "CounterData");
        instance = this;
        config = new Config("counters");
        loadAll();
    }

    public static CounterStore getInstance() {
        return CounterStore.instance;
    }

    /**
     * 名前を用いてCounterDataを取得します。
     * @param name CounterDataの名前
     * @return 対応するCounterData。なければnull
     */
    public @Nullable CounterData get(String name) {
        return counters.get(name);
    }

    /**
     * 始点座標を用いてCounterDataを取得します。
     * @return 対応するCounterData。なければnull
     */
    public @Nullable CounterData getByLocation1(Location location) {
        return location1Index.get(location);
    }

    /**
     * 終点座標を用いてCounterDataを取得します。
     * @return 対応するCounterData。なければnull
     */
    public @Nullable CounterData getByLocation2(Location location) {
        return location2Index.get(location);
    }

    /**
     * カウンターの一覧を取得します。
     * @return カウンターの一覧。
     */
    public List<CounterData> getCounters() {
        return counters.values().stream().toList();
    }

    /**
     * CounterDataを追加します。
     * @param data 追加するデータ
     * @throws IOException 保存に失敗した
     */
    public void add(CounterData data) throws IOException {
        counters.put(data.getName(), data);
        addToIndex(data);
        update(data);
    }

    /**
     * CounterDataを削除します。
     * @param data 削除するデータ
     * @throws IOException 保存に失敗した
     */
    public void remove(CounterData data) throws IOException {
        remove(data.getName());
    }

    public void update(CounterData data) throws IOException {
        if (!counters.containsKey(data.getName())) throw new IllegalArgumentException();
        config.getConf().set(data.getName(), data);
        config.save();
    }

    /**
     * CounterDataを削除します。
     * @param name 削除するデータの名前
     * @throws IOException 保存に失敗した
     */
    public void remove(String name) throws IOException {
        removeFromIndex(get(name));
        counters.remove(name);
        config.getConf().set(name, null);
        config.save();
    }

    /**
     * 全プレイヤーのデイリーイベントプレイ履歴を削除します。
     * @throws IOException 保存に失敗した
     */
    public void resetAllPlayersPlayedLog() throws IOException {
        final var pstore = PlayerStore.getInstance();
        pstore.openAll()
            .forEach(record -> record.delete(PlayerDataKey.PLAYED_COUNTER, false));
        pstore.save();
    }

    /**
     * インデックスにカウンターデータを追加する
     */
    private void addToIndex(CounterData data) {
        location1Index.put(data.getLocation1(), data);
        location2Index.put(data.getLocation2(), data);
    }

    /**
     * インデックスからカウンターデータを削除する
     */
    private void removeFromIndex(CounterData data) {
        location1Index.remove(data.getLocation1());
        location2Index.remove(data.getLocation2());
    }

    private void loadAll() {
        final var yml = config.getConf();
        yml.getKeys(false).forEach(name -> {
            final var counter = yml.getObject(name, CounterData.class);
            counters.put(name, counter);
            addToIndex(counter);
        });
    }

    private static CounterStore instance;

    /** counters.yml */
    private Config config;

    /** カウンターデータのマップ */
    private final Map<String, CounterData> counters = new HashMap<>();

    /** 始点座標による索引 */
    private final Map<Location, CounterData> location1Index = new HashMap<>();
    /** 終点座標による索引 */
    private final Map<Location, CounterData> location2Index = new HashMap<>();
}

File: src/main/java/work/xeltica/craft/core/stores/CounterStore.kt
package work.xeltica.craft.core.stores

import org.bukkit.Location
import org.bukkit.configuration.serialization.ConfigurationSerialization
import work.xeltica.craft.core.api.Config
import work.xeltica.craft.core.models.CounterData
import kotlin.Throws
import java.io.IOException
import work.xeltica.craft.core.models.PlayerRecord
import work.xeltica.craft.core.models.PlayerDataKey
import java.util.HashMap
import java.util.function.Consumer

/**
 * 時間計測カウンターの情報を管理します。
 */
class CounterStore {
    /**
     * 名前を用いてCounterDataを取得します。
     * @param name CounterDataの名前
     * @return 対応するCounterData。なければnull
     */
    operator fun get(name: String): CounterData? {
        return counters[name]
    }

    /**
     * 始点座標を用いてCounterDataを取得します。
     * @return 対応するCounterData。なければnull
     */
    fun getByLocation1(location: Location): CounterData? {
        return location1Index[location]
    }

    /**
     * 終点座標を用いてCounterDataを取得します。
     * @return 対応するCounterData。なければnull
     */
    fun getByLocation2(location: Location): CounterData? {
        return location2Index[location]
    }

    /**
     * カウンターの一覧を取得します。
     * @return カウンターの一覧。
     */
    fun getCounters(): List<CounterData> {
        return counters.values.stream().toList()
    }

    /**
     * CounterDataを追加します。
     * @param data 追加するデータ
     * @throws IOException 保存に失敗した
     */
    @Throws(IOException::class)
    fun add(data: CounterData) {
        counters[data.name] = data
        addToIndex(data)
        update(data)
    }

    /**
     * CounterDataを削除します。
     * @param data 削除するデータ
     * @throws IOException 保存に失敗した
     */
    @Throws(IOException::class)
    fun remove(data: CounterData) {
        remove(data.name)
    }

    @Throws(IOException::class)
    fun update(data: CounterData) {
        require(counters.containsKey(data.name))
        config.conf[data.name] = data
        config.save()
    }

    /**
     * CounterDataを削除します。
     * @param name 削除するデータの名前
     * @throws IOException 保存に失敗した
     */
    @Throws(IOException::class)
    fun remove(name: String) {
        removeFromIndex(get(name))
        counters.remove(name)
        config.conf[name] = null
        config.save()
    }

    /**
     * 全プレイヤーのデイリーイベントプレイ履歴を削除します。
     * @throws IOException 保存に失敗した
     */
    @Throws(IOException::class)
    fun resetAllPlayersPlayedLog() {
        val pstore = PlayerStore.instance
        pstore!!.openAll()
            .forEach(Consumer { record: PlayerRecord -> record.delete(PlayerDataKey.PLAYED_COUNTER, false) })
        pstore.save()
    }

    /**
     * インデックスにカウンターデータを追加する
     */
    private fun addToIndex(data: CounterData) {
        location1Index[data.location1] = data
        location2Index[data.location2] = data
    }

    /**
     * インデックスからカウンターデータを削除する
     */
    private fun removeFromIndex(data: CounterData?) {
        location1Index.remove(data!!.location1)
        location2Index.remove(data.location2)
    }

    private fun loadAll() {
        val yml = config.conf
        yml.getKeys(false).forEach(Consumer { name: String ->
            val counter = yml.getObject(name, CounterData::class.java)!!
            counters[name] = counter
            addToIndex(counter)
        })
    }

    /** counters.yml  */
    private val config: Config

    /** カウンターデータのマップ  */
    private val counters: MutableMap<String, CounterData> = HashMap()

    /** 始点座標による索引  */
    private val location1Index: MutableMap<Location, CounterData> = HashMap()

    /** 終点座標による索引  */
    private val location2Index: MutableMap<Location, CounterData> = HashMap()

    init {
        ConfigurationSerialization.registerClass(CounterData::class.java, "CounterData")
        instance = this
        config = Config("counters")
        loadAll()
    }

    companion object {
        @JvmStatic
        lateinit var instance: CounterStore
            private set
    }
}